File: src/network/packets/updateLeaderBoard.js
import Packet from '../packet.js';
import Writer from '../utils/writer.js';
import Point from "../../ui/objects/point";
import Viewer from "../../ui/objects/viewer";
import {convertIntColorToHex} from "../../ui/utils";


class UpdateLeaderBoardPacket extends Packet {

    constructor() {
        super();
        this.packetId = 1013;
        this.players = [];

    }


    // Handel Server Response
    parsePacket() {
        const reader = this.reader;
        const playersLength = reader.readInt2();
        for (let i = 0; i < playersLength; i++) {
            const name = reader.readString();
            const id = reader.readInt4();
            const numberOfKills = reader.readInt4();
            const playerColor = convertIntColorToHex(reader.readInt4());
            const score = Math.round(reader.readFloatFromBytes() * 100) / 100;

            if (window.myPlayer && window.myPlayer.id === id){
                window.myPlayer.score = score;
                window.myPlayer.kills = numberOfKills;
                window.myPlayer.rank = i + 1;
            }



            this.players.push(
                {
                    name: name,
                    playerColor: playerColor,
                    score,
                    id,
                    numberOfKills,
                }
            );
        }



    }

    finalize() {
        const writer = new Writer(this.packetId);
        return writer.finalize();
    }


    handleReceivedPacket(client) {
        window.leaderboard.updateLeaderboard(this.players);
    }
}

export default UpdateLeaderBoardPacket;

File: gameserver/network/packets/update_leader_board.py
import time

from gameserver.game.viewer import Viewer
from gameserver.network.packet import Packet
from gameserver.network.utils.writer import Writer


class UpdateLeaderBoardPacket(Packet):
    PACKET_ID = 1013

    def __init__(self,players):
        super().__init__()
        self.players = players



    def parse_packet(self):
        pass



    def handle_packet(self, client):
        """On Received Ready Packet"""
        pass




    def finalize(self):
        writer = Writer(self.packet_id)
        writer.write_int_in_bytes(len(self.players), 2)
        for player in self.players:
            writer.write_string(player.name)
            writer.write_int_in_bytes(player.player_id, 4)
            writer.write_int_in_bytes(player.numberOfKills, 4)
            writer.write_int_in_bytes(player.color_pattern, 4)
            percentage = player.occupied_percentage
            rounded = round(percentage, 2)
            writer.write_float_in_bytes(rounded, 4)

        return writer.finalize()
